core-web/apps/dotcms-ui/src/app/api/services/dot-templates/dot-templates.service.ts

```diff
@@ -185,4 +185,18 @@ export class DotTemplatesService {
             })
         );
     }
+
+    private addContainerInfo(template: DotTemplate): DotTemplate {
+        if (template?.layout?.body?.rows)
+            for (const row of template.layout.body.rows) {
+                if (row.columns)
+                    for (const column of row.columns) {
+                        for (const container of column.containers) {
+                            container.title = template.containers[container.identifier].title;
+                        }
+                    }
+            }
```

Comment:
This can be really slow. can we find another way to do this?

Reply:
We could do something like this:

```
private addContainerInfo(template: DotTemplate): DotTemplate {
        if (template?.layout?.body?.rows) {
            const updatedRows = template.layout.body.rows.map((row) => {
                return row.columns
                    ? {
                          ...row,
                          columns: row.columns.map((column) =>
                              column.containers
                                  ? {
                                        ...column,
                                        containers: column.containers.map((container) => ({
                                            ...container,
                                            title: template.containers[container.identifier].title
                                        }))
                                    }
                                  : column
                          )
                      }
                    : row;
            });

            return {
                ...template,
                layout: {
                    ...template.layout,
                    body: {
                        ...template.layout.body,
                        rows: updatedRows
                    }
                }
            };
        }

        return template;
    }
```

but I feel like it may sacrifice readability. What do you think?

Reply:
To improve the readability of this code we can use the guard approach, instead of doing a ternary we can do:

```javascript
if(!row.columns)
  return row;
```

The same inside the columns map.

Also you can use destructuring:

```javascript
//...
 template.layout.body.rows.map((row) => { 
     const { columns } = row;
     //...
 }
```

You can also do this inside the columns map

It increase readablity, but this way we are not doing 3 nested loops for just adding a name.

Reply:
@zJaaal Are you completely sure about performance being far worse just because of the loops?
I know that it can be bad when we have cases where we can avoid iterating some entities, however in this scenario we need to add a name for each of the containers. Did some runs of this PerformanceTester, and the original solution performed better in each run. Wanted to know your opinion on it.

```
interface DotContainer {
    identifier: string;
    title: string;
}

interface DotColumn {
    containers: DotContainer[];
}

interface DotRow {
    columns: DotColumn[];
}

interface DotBody {
    rows: DotRow[];
}

interface DotLayout {
    body: DotBody;
}

interface DotTemplate {
    layout: DotLayout;
    containers: Record<string, DotContainer>;
}

class PerformanceTester {
    private addContainerInfoOriginal(template: DotTemplate): DotTemplate {
        if (template?.layout?.body?.rows) {
            for (const row of template.layout.body.rows) {
                if (row.columns) {
                    for (const column of row.columns) {
                        for (const container of column.containers) {
                            container.title = template.containers[container.identifier].title;
                        }
                    }
                }
            }
        }

        return template;
    }

    private addContainerInfoModified(template: DotTemplate): DotTemplate {
        if (template?.layout?.body?.rows) {
            return {
                ...template,
                layout: {
                    ...template.layout,
                    body: {
                        ...template.layout.body,
                        rows: template.layout.body.rows.map(row => {
                            if (row.columns) {
                                return {
                                    ...row,
                                    columns: row.columns.map(column => {
                                        if (column.containers) {
                                            return {
                                                ...column,
                                                containers: column.containers.map(container => ({
                                                    ...container,
                                                    title: template.containers[container.identifier].title
                                                }))
                                            };
                                        }
                                        return column;
                                    })
                                };
                            }
                            return row;
                        })
                    }
                }
            };
        }

        return template;
    }

    private measurePerformance(fn: () => void): number {
        const startTime = performance.now();
        fn();
        const endTime = performance.now();
        return endTime - startTime;
    }

    public runPerformanceTest(): void {
        // Create mock data
        const container1: DotContainer = { identifier: '1', title: 'Container 1' };
        const container2: DotContainer = { identifier: '2', title: 'Container 2' };
        const container3: DotContainer = { identifier: '3', title: 'Container 3' };
        const container4: DotContainer = { identifier: '4', title: 'Container 4' };

        const template: DotTemplate = {
            layout: {
                body: {
                    rows: [
                        {
                            columns: [
                                {
                                    containers: [container1, container2]
                                }
                            ]
                        },
                        {
                            columns: [
                                {
                                    containers: [container3, container4]
                                }
                            ]
                        }
                    ]
                }
            },
            containers: {
                '1': { identifier: '1', title: 'Updated Container 1' },
                '2': { identifier: '2', title: 'Updated Container 2' },
                '3': { identifier: '3', title: 'Updated Container 3' },
                '4': { identifier: '4', title: 'Updated Container 4' }
            }
        };

        // Measure performance
        const originalTime = this.measurePerformance(() => this.addContainerInfoOriginal(template));
        const modifiedTime = this.measurePerformance(() => this.addContainerInfoModified(template));

        // Output results
        console.log(`Execution time (Original): ${originalTime} milliseconds`);
        console.log(`Execution time (Modified): ${modifiedTime} milliseconds`);
    }
}

// Usage
const tester = new PerformanceTester();
tester.runPerformanceTest();
```

Reply:
I think that the other solution is the same but with JS methods, is still 3 nested loops. Could be slower because it needs to access to JS methods, which is slower than just execute a for. 

We can stick to the first approach which is more readable, but maybe we can find another way to do this. Like getting this from the backend or something. We can discuss that in a meeting and see what the team thinks about it

Reply:
Was just curious, ended up going with your approach. Think it would be nice getting it from the backend, though. Agree that it would be an improvement 👍